### PR TITLE
feat(nest): support nest 8

### PR DESCRIPTION
**Pull request by gperdomor**
## Current Behavior
Now Nest 7 is in use

## Expected Behavior
Use Nest 8

## Related Issue(s)

Fixes #

### CODE REVIEW

File: packages/nest/src/utils/versions.ts
Comment (Ionaru) [inline packages/nest/src/utils/versions.ts:6]: How will this interact with Angular (https://github.com/nrwl/nx/blob/master/packages/angular/src/utils/versions.ts#L5) that still requires RxJS 6? I'm asking this because Nx offers an Angular + NestJS workspace by default.

Angular support for RxJS 7 is pending and is expected to arrive in the next major version, ref: https://github.com/angular/angular/issues/41897

Comment (gperdomor) [reply packages/nest/src/utils/versions.ts:6]: Yes, indeed this is a problem, with Nest + Next/React should not be any problem thought

**Comment by AgentEnder** [comment]
Hey @gperdomor. This PR looks good overall, but I am marking it as blocked due to the RxJs incompatibility with angular currently. When that is resolved, I will reevaluate this PR. Based on the angular issue linked above, I would estimate that to be after Ng13 is released.

**Comment by AliYusuf95** [comment]
I think having pakage.json per-app https://github.com/nrwl/nx/issues/1777 would be the optimal solution for these cases.
I've been in this situation many times... it's very annoying.

**Comment by gperdomor** [comment]
> Hey @gperdomor. This PR looks good overall, but I am marking it as blocked due to the RxJs incompatibility with angular currently. When that is resolved, I will reevaluate this PR. Based on the angular issue linked above, I would estimate that to be after Ng13 is released.

@AgentEnder That's ok by me... anyway I'm in NX 12.5.8 and I have an app with Nest 8 and seems to work well with express adapter but is not served with the FastifyAdapter, I'm wondering if you can help me to find why... :(

File: packages/nest/src/utils/versions.ts
Comment (Nightbr) [reply packages/nest/src/utils/versions.ts:6]: Yep, confirmed that I upgraded my workspace to Nestjs v8 with RxJS v7 (using React App). Perhaps it could be an opt-in option to set Nestjs v8 (with RxJS v7).

File: packages/nest/src/generators/application/files/main.ts__tmpl__
Comment (gperdomor) [reply packages/nest/src/generators/application/files/main.ts__tmpl__:14]: Honestly nothing special here, I just want to sync with the official template, but this can be reverted :)

**Comment by AgentEnder** [comment]
@gperdomor If you resolve conflicts, this should be able to be evaluated at an opt-in level with angular 12.2 support

**Comment by gperdomor** [comment]
@AgentEnder done 😄

File: packages/nest/src/generators/application/files/main.ts__tmpl__
Comment (nartc) [reply packages/nest/src/generators/application/files/main.ts__tmpl__:14]: nitpick/suggestion: This `globalPrefix` is actually being used to setup the `proxy.conf` for a `frontendProject` if it exists. Per discussion with @AgentEnder , there also might be changes to how we generate/use `globalPrefix` as well. So as of the moment, I'd like to keep the `globalPrefix` (or the tmpl file) as is.

Comment (gperdomor) [reply packages/nest/src/generators/application/files/main.ts__tmpl__:14]: ok I will revert this file :D

Comment (gperdomor) [reply packages/nest/src/generators/application/files/main.ts__tmpl__:14]: @nartc done

**Comment by gperdomor** [comment]
@AgentEnder rebased with master 👍🏻

**Comment by AgentEnder** [comment]
@gperdomor Can you replicate the failing e2e test locally? I'm hoping we can get this merged ASAP, but its currently still blocked by Angular.

**Comment by gperdomor** [comment]
> @gperdomor Can you replicate the failing e2e test locally? I'm hoping we can get this merged ASAP, but its currently still blocked by Angular.

@AgentEnder Finally passes

**Comment by gperdomor** [comment]
ping @AgentEnder :)

**Comment by AgentEnder** [comment]
Hey @gperdomor. We discussed this some more internally, if you remove the changes that add getUrl() in main.ts we can support either nest 7 or nest 8 inside of a repo.

The next step towards that would be updating the migration logic to only bump Nest + RxJS if Ngrx is not present in the workspace. This is needed as NgRx doesn't support RxJS 7 yet, and would let users who can update do so optionally for now. 

I'm going to tag @jaysoo and @nartc here, as this process was the result of our discussions.

**Comment by nartc** [comment]
Thank you @AgentEnder for leaving the comment.

Huge thank you @gperdomor for keeping this PR up to date. We do realize the pain point so we haven't been very attentive (I'm going to use Nx Conf as an excuse) but we do track this PR movements via email notification.

As Craigory mentioned, we discussed this internally and decided to come up with a smarter migration logic. In summary, we'd like:
- NestJS 7 stays on NestJS 7 and RxJS 6 if NgRX is not a dependency
- NestJS 7 and RxJS will be bumped to 8 and 7 respectively.
- New generated workspace will have NestJS 8 and RxJS 7 installed instead

Again, thank you for your effort. It is very much appreciated.

**Comment by gperdomor** [comment]
@nartc @AgentEnder I just updated the PR, but honestly, I do not know if it is the correct way to do what you ask. I do not know how smart migrations works but I take an old migration to Nest 7 as reference.

**Comment by AgentEnder** [comment]
@nartc Do you care to look over this and test the migration on a few repos? It LGTM, but I've not had time to check it out against any tests.

File: packages/nest/src/migrations/update-13-2-0/update-13-2-0.ts
Comment (nartc) [inline packages/nest/src/migrations/update-13-2-0/update-13-2-0.ts:1]: suggestion: let's switch to use `nrwl/devkit` instead. 
image

If you have this box checked in the PR, I can push to switch to `nrwl/devkit`

**Review by nartc — COMMENTED** [review]
Hi @gperdomor , sorry for dragging this PR for too long, and thank you for the effort thus far.

We've been going back and forth internally on how to go ahead with this to be the most straightforward for the consumers.

The logic has been changed just slightly from what we currently have in this PR. 
image

We probably want to do the same thing for `nrwl/nest:app` generator as well.

cc @AgentEnder if I miss anything, please feel free to add

Thanks again.

File: packages/nest/src/migrations/update-13-2-0/update-13-2-0.ts
Comment (gperdomor) [reply packages/nest/src/migrations/update-13-2-0/update-13-2-0.ts]: Is checked 😄 
Screen Shot 2021-10-30 at 12 05 28 AM

**Comment by gperdomor** [comment]
@nartc @AgentEnder PR updated with the new migration and generator logic

File: packages/nest/migrations.json
Comment (gperdomor) [reply packages/nest/migrations.json]: deleted 😄

**Comment by nartc** [comment]
Overall LGTM, just a couple questions and suggestions

File: packages/nest/src/generators/application/files/main.ts__tmpl__
Comment (gperdomor) [reply packages/nest/src/generators/application/files/main.ts__tmpl__:17]: Nothing fancy here, it's only to not mix async/await with callbacks, seems a little cleaner IMO

File: packages/nest/src/migrations/update-13-2-0/update-to-nest-8.ts
Comment (gperdomor) [reply packages/nest/src/migrations/update-13-2-0/update-to-nest-8.ts:24]: updated 👍🏻

File: packages/nest/src/generators/application/files/main.ts__tmpl__
Comment (nartc) [reply packages/nest/src/generators/application/files/main.ts__tmpl__:17]: Gotcha

File: packages/nest/package.json
Comment (nartc) [reply packages/nest/package.json:37]: `semver` is `@nrwl/workspace` dependency so technically the consumers should have `semver` installed in their project already. 

That said, you can test this locally too:

- Create an nx workspace locally
- Enable and start local registry 
- Run nx-release with --local flag
- Try running migrate in your workspace

Comment (nartc) [reply packages/nest/package.json:37]: @AgentEnder pointed out a good point about `semver` and CI Dep Check. You might want to add `semver` to `missings.ts` for `nest` in `IGNORE_MATCHES`

PS: Look at `angular` in `IGNORE_MATCHES`

Comment (gperdomor) [reply packages/nest/package.json:37]: Tests:

- Empty Workspace: For empty workspace I run `nx migrate latest` command to update to the 13.2.0 version from local registry, I generate a new nest application and nest 8 was used

- Nest Workspace: In a new nest workspace, I run the `nx migrate latest`command, a migration file was created, then run `nx migrate --run-migrations` and throws with `Schematic "update-13-2-0" cannot resolve the factory.`

```
>  NX  Running migrations from 'migrations.json'

Running migration update-13-2-0
Schematic "update-13-2-0" cannot resolve the factory.
/Users/gperdomor/Downloads/only-nest/node_modules/yargs/build/lib/yargs.js:1132
                throw err;
                ^

Error: Command failed: /var/folders/wz/pcsm158j3bj627wy88cvrr6m0000gn/T/tmp-25427-Jiv5tG72cHuq/node_modules/.bin/tao migrate --run-migrations
    at checkExecSyncError (node:child_process:826:11)
    at Object.execSync (node:child_process:900:15)
    at Object.handler (/Users/gperdomor/Downloads/only-nest/node_modules/@nrwl/workspace/src/command-line/nx-commands.js:114:25)
    at Object.runCommand (/Users/gperdomor/Downloads/only-nest/node_modules/yargs/build/lib/command.js:196:48)
    at Object.parseArgs [as _parseArgs] (/Users/gperdomor/Downloads/only-nest/node_modules/yargs/build/lib/yargs.js:1043:55)
    at Object.get [as argv] (/Users/gperdomor/Downloads/only-nest/node_modules/yargs/build/lib/yargs.js:986:25)
    at Object.initLocal (/Users/gperdomor/Downloads/only-nest/node_modules/@nrwl/cli/lib/init-local.js:28:79)
    at Object.<anonymous> (/Users/gperdomor/Downloads/only-nest/node_modules/@nrwl/cli/bin/nx.js:43:18)
    at Module._compile (/Users/gperdomor/.nvm/versions/node/v16.13.0/lib/node_modules/nx/node_modules/v8-compile-cache/v8-compile-cache.js:192:30)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1153:10) {
  status: 1,
  signal: null,
  output: [ null, null, null ],
  pid: 25499,
  stdout: null,
  stderr: null
}
```

Same error for Angular+Nest workspaces

Comment (nartc) [reply packages/nest/package.json:37]: Lemme pull this PR down and try locally. Thank you for your time

Comment (gperdomor) [reply packages/nest/package.json]: I solve the factory problem, pushing changes now... But I have other problem, calling `addDependenciesToPackageJson` in the line 63 is not updating the packages to the new versions...

Comment (nartc) [reply packages/nest/package.json]: @gperdomor `addDependenciesToPackageJson` does not **update** the packages. You'd need to call `updateJson` manually

Comment (gperdomor) [reply packages/nest/package.json]: PR updated, the generator and migration works properly now, at least in all my tests

**Comment by nartc** [comment]
@gperdomor  Thank you for the effort. Doing the last round of testing locally after a meetup. Will keep you all updated

File: packages/nest/src/generators/init/lib/add-dependencies.ts
Comment (nartc) [inline packages/nest/src/generators/init/lib/add-dependencies.ts:23]: suggestion/nitpick: let's use `angular/core` instead of `angular/common`. Though it doesn't really matter in this case, but we feel like `angular/core` is a tad better. That's why it's a "nitpick" ^^

File: packages/nest/src/migrations/update-13-2-0/update-to-nest-8.ts
Comment (nartc) [inline packages/nest/src/migrations/update-13-2-0/update-to-nest-8.ts:53]: nitpick: same as above about `angular/common`

suggestion: let's clean up the `else` blocks in this check. If you return a value, you don't need `else`.

```suggestion
if (json.dependencies['@angular/common']) {
    const rxjs = checkAndCleanWithSemver('rxjs', json.dependencies['rxjs']);
    if (satisfies(rxjs, rxjsVersion7)) {
      return true;
    }

    const { Confirm } = require('enquirer');
    const prompt = new Confirm({
      name: 'question',
      message: 'Do you want to update to RxJS 7 + Nest 8?',
      initial: true,
    });

    return await prompt.run();
  }

  return true;
```

File: packages/nest/src/utils/versions.ts
Comment (nartc) [inline packages/nest/src/utils/versions.ts:7]: suggestion/question: this is more of a discussion.

Since we set `@nestjs/schematics` to v8 in `nrwl/nest` `package.json`, should we always force `@nestjs/schematics` to v8 regardless?

I tested and `@nestjs/schematics` does not depend on `rxjs` or anything. The problem with `@nestjs/schematics` staying on v7 is when Angular 13 is part of the migration of the next Nx release, then `@nestjs/schematics@7` will fail because of failed schema validation (due to Angular 13 forcing the `$id` in the schematics schema)

**Review by nartc — COMMENTED** [review]
Last couple nitpicks and suggestions

File: packages/nest/src/utils/versions.ts
Comment (gperdomor) [reply packages/nest/src/utils/versions.ts]: agree, i check the @nestjs/schematics dependencies and it seems to be safe to use v8 in both cases

**Review by nartc — APPROVED** [review]
LGTM! Thank you so much for this PR!!

**Comment by gperdomor** [comment]
> LGTM! Thank you so much for this PR!!

I'm happy to help 😄 👍🏻

File: packages/nest/src/migrations/update-13-2-0/update-to-nest-8.ts
Comment (nartc) [inline packages/nest/src/migrations/update-13-2-0/update-to-nest-8.ts:29]: issue: @gperdomor apparently this doesn't work, consumers still need to run install package manually. I think we can log something here instead. Check out the `storybook` migrations

Comment (nartc) [reply packages/nest/src/migrations/update-13-2-0/update-to-nest-8.ts]: Pushed the change

**Comment by nartc** [comment]
@gperdomor once again, thank you. I had one last issue with the PR (about `installPackagesTask()`) but I pushed the change myself. As soon as this thing turns green, we're good to merge

File: packages/nest/src/migrations/update-13-2-0/update-to-nest-8.ts
Comment (gperdomor) [reply packages/nest/src/migrations/update-13-2-0/update-to-nest-8.ts]: I see you already fixed that 👍🏻

**Pull request merged**
